### ISSUE TITLE
Use macros to define a constant and declare a data table entry at the same time

### ISSUE THREAD

**Issue by Rangi42** [pret/pokecrystal#748]
For general-purpose constants we have the `const` macro. These often correspond to many different data tables, as with Pokémon constants, item constants, move constants, etc. They also get directly used in code with `cp`, `ld`, etc. These are fine as-is. (Although constants that correspond to struct offsets could use the `RS` commands; see issue #739.)

Other times we have a single pointer table, and a single set of "scripts" that list one-byte indexes into that table. For these, we define one macro to declare table entries, and one to declare byte indexes as a difference from the start of the table. For example: `add_mapsetup` and `mapsetup`; `add_tradeanim` and `tradeanim`; `add_setup_command` and `pokeanim`; `add_stdscript` and `callstd`/`jumpstd`; `add_predef` and `predef`/`predef_jump`; etc. These are also fine as-is: they don't need any redundant declarations, and the "scripts" use the same names as the actual pointer entries, not similar-looking all-caps or all-lowercase constants.

Other times we have similar pointer tables and one-byte offsets, but the offsets are used very frequently, and sometimes take parameters as extra bytes. For these, we define individual macros, since what's most important is for the scripts to be readable and editable. For example, map event scripts, text commands, battle animations, audio commands, basically everything in macros/scripts/.

Finally, we have local pointer tables (usually jumptables) where the one-byte indexes generally get used in the code, not just declared in `db` scripts. Often there will be different routines that load new jumptable indexes into `a` or `[wJumptableIndex]`, or might `cp` some of them. The current solution for those is to declare constants at the top of the file which correspond to the jumptable: see engine/pokegear/pokegear.asm, engine/debug/debug_room.asm, engine/pokedex/pokedex.asm, and others. This is awkward since you have to maintain two lists and two sets of names, and many small jumptables haven't bothered to define index constants yet.

I think a simpler solution for them would be a single macro that defines the constant and declares the pointer at the same time.

```
const_dw: MACRO
    const \1
    dw \2
ENDM
```

So code like this:

```
	const_def 1
	const PINK_PAGE  ; 1
	const GREEN_PAGE ; 2
	const BLUE_PAGE  ; 3
NUM_STAT_PAGES EQU const_value - 1

...

.Jumptable:
; entries correspond to *_PAGE constants
	dw LoadPinkPage
	dw LoadGreenPage
	dw LoadBluePage
```

could be rewritten as:

```
.Jumptable:
	const_def
	const_dw PINK_PAGE,  LoadPinkPage
	const_dw GREEN_PAGE, LoadGreenPage
	const_dw BLUE_PAGE,  LoadBluePage
NUM_STAT_PAGES EQU const_value - 1
```

As of rgbds 0.4.1 the constants can even be used in code that's above where they're declared.

**Comment by Rangi42** [comment]
@mid-kid What do you think? And could any improvements be bade to the other kinds of constants?

**Comment by mid-kid** [comment]
> As of rgbds 0.4.1 the constants can even be used in code that's above where they're declared.

Is this true if they're not `EXPORT`ed?

I don't like `const_dw` as a name, but can't really think of proper alternatives. Something with "jumptable"? like `jump_def`? Considering this is the only way in which they're going to be used, something along those lines seems appropriate, imo.

Also, possibly kind of unrelated to this, but things like `special`, `predef` and friends should probably `EXPORT` the indexes into the table, instead of their label and calculating it at link time. That would simplify the macros and be a more sane implementation.

We're essentially defining four kinds of tables here:
- Offsets with arguments (commands) get their own macros in macros/scripts
- Offsets with no arguments, that are only used as a direct `call` reference get a macro so you can treat them as regular function calls
- Offsets with no arguments, of which their indexes are passed around and modified by code (not just fetched from a table to index a jumptable) and are only used locally (preferably in the same file), get a constant defined at the only jumptable that uses them.
- Offsets with no arguments, that are used to index multiple tables get separate constant definitions in `constants/`

The big, big question here, is in what category do things like `SPRITE_ANIM_INDEX_`/`SEQ_`/`FRAMESET_`/etc belong? Currently they're treated like the fourth category but I'd say they belong to the third or maybe even the second if the description is slightly modified to include atomic data definitions (i.e. "only used as a direct reference to data or code"). And I'm sure they're not the only ones.

**Comment by Rangi42** [comment]
If the uses really are all for jumptables (probably), then perhaps `jump_dest` or `jump_target`. `jump_def` would also be fine, it emphasizes the defining of a new constant.

The issue with `special`, `predef`, `stdscript`, etc exporting constants would be that it creates a second name to keep track of. Possibilities:

- Explicitly define a new name, same as how the proposed `const_dw` or whatever would work, like `add_stdscript tv, TVScript`. Then `callstd tv` would just have to do `db \1` internally. The issue is, this creates a second name to keep track of, so you can't look at the `callstd` invocation and know what script it's for.
- Implicitly add a prefix/suffix at the definition and the invocation, like how it currently does with the label name. So `add_stdscript TVScript` would define `\1StdScript` as an incrementing constant instead of a label on its table entry, and `callstd TVScript` would do `db \1StdScript` instead of `dw (\1StdScript - StdScripts) / 3`. This is probably what you meant; I like how it would make all uses of the constant simpler, no need to know the table's name and entry width.

I'd be okay with changing some non-jumptable data tables to define their own constants inline, on a case-by-case basis. The sprite animation system might be a good candidate for that. E.g.:

```
sprite_anim_index: MACRO
	const \1
EXPORT \1
	db \2, \3, \4
ENDM

SpriteAnimSeqData:
	; frameset, sequence, tile
	sprite_anim_index SPRITE_ANIM_INDEX_PARTY_MON, SPRITE_ANIM_FRAMESET_PARTY_MON, SPRITE_ANIM_SEQ_PARTY_MON, $00
```

```
DoAnimFrame:
	...

.Jumptable:
	const_dw SPRITE_ANIM_SEQ_NULL,      .Null
	const_dw SPRITE_ANIM_SEQ_PARTY_MON, .PartyMon
```

And likewise for `SPRITE_ANIM_FRAMESET_*` and `SPRITE_ANIM_OAMSET_*`.

Then again, tables like `DoAnimFrame.Jumptable` are buried in engine/ code, yet their corresponding constants are used by many other files, and the core "add a new sprite animation" task requires you to define a new one. For which IMO it's useful to make them easily discoverable, even if means having two parallel lists. Currently all the relevant constants are listed in sprite_anim_constants.asm with comments saying which data or engine tables also need updating.

We could always start with local jumptables, see how it affects usability, and then consider refactoring other tables.

**Comment by mid-kid** [comment]
Yeah, I meant the second bullet point. We already have these symbols - just change their values. I don't think much about the way these macros are used should be changed.

As for the sprite anims, I mostly mentioned it as something worth looking into/considering in the future. I remember having to constantly tab between 5 files to add a new object with minimal behavior, and I think some things like oamsets are pretty intertwined with framesets and never used standalone, for which I'd rather not need to define a global constant. I could be wrong though, it's been a while.

**Comment by Rangi42** [comment]
Yeah, `FRAMESET`s are the only content that references `OAMSET`s. On the other hand, `FRAMESET`s, `SEQ`s, and `INDEX`es are all used in engine code, so may as well just declare all four constant types sequentially together.

(Same goes for `ANIM_OBJ_*`, `BATTLEANIMFRAMESET_*`, and `BATTLEANIMOAMSET_*`, and the related `BATTLEANIMFUNC_*`, although these and their pokered equivalents are both poorly labeled/understood. Getting them documented might reveal some other ways the two animation systems could be simplified together.)

**Comment by Kroc** [comment]
These could be called `enum`s

```
.Jumptable:
	enum_def
	enum PINK_PAGE,  LoadPinkPage
	enum GREEN_PAGE, LoadGreenPage
	enum BLUE_PAGE,  LoadBluePage
```

**Comment by Rangi42** [comment]
That would be ambiguous with the now-deleted `enum`.

**Comment by Rangi42** [comment]
Combining this proposal with #739, we could use https://github.com/ISSOtm/rgbds-structs or something similar to simultaneously define struct offset constants and a macro to instantiate the struct.

One difference from rgbds-structs would be that we use `UPPER_CASE` for offset constants and `TitleCase` for labels. So where we currently have this:

```
box_struct: MACRO
\1Species::        db
\1Item::           db
\1Moves::          ds NUM_MOVES
\1ID::             dw
\1Exp::            ds 3
...
ENDM
```

```
MON_SPECIES            EQUS "(wPartyMon1Species - wPartyMon1)"
MON_ITEM               EQUS "(wPartyMon1Item - wPartyMon1)"
MON_MOVES              EQUS "(wPartyMon1Moves - wPartyMon1)"
MON_ID                 EQUS "(wPartyMon1ID - wPartyMon1)"
MON_EXP                EQUS "(wPartyMon1Exp - wPartyMon1)"
...
```

We could instead have something like this:

```
	struct box_struct
	sb MON_SPECIES, Species
	sb MON_ITEM,    Item
	sb MON_MOVES,   Moves, 4
	sw MON_ID,      ID
	sb MON_EXP,     Exp, 3
	...
	end_struct
```

Or with an implicit constant prefix:

```
	struct MON, box_struct
	sb SPECIES, Species
	sb ITEM,    Item
	sb MOVES,   Moves, 4
	sw ID,      ID
	sb EXP,     Exp, 3
	...
	end_struct
```

**Comment by aaaaaa123456789** [comment]
Honestly, I'd argue against using rgbds-structs (or anything like that) purely from a readability standpoint. I've had to direct people to macro definitions countless times, and I wouldn't want to send anyone to that file — it's hard to read even for experienced developers (and for a good reason).

`MON_ITEM EQUS "(wPartyMon1Item - wPartyMon1)"` is something you can understand right away once you read it, and you can easily see _why_ that constant refers to a struct offset. Struct definitions are equally easy to read — the macros are obvious once you know how macros work. Having readable macros is important, because people who haven't been working on the codebase for years can't easily abstract them away without understanding them first, and when those people aren't particularly experienced with RGBDS, this leads them to have a hard time understanding the code. (As a simple test, I challenge you to redirect someone who isn't too experienced to pokecrystal16's [indirection table macros](https://github.com/aaaaaa123456789/pokecrystal16/blob/base-components/macros/indirection.asm), which aren't particularly complex, and see if they can understand them.)

**Comment by Rangi42** [comment]
`MON_ITEM EQUS "(wPartyMon1Item - wPartyMon1)"` requires choosing one usage of the struct to be the definitive one for its offsets.

Issue #739 only suggests changing how struct offsets are defined, not like this issue which suggests combining the definitions of the offset constant and the macro label. For example, this:

```
; party_struct members (see macros/wram.asm)
    rsreset
MON_SPECIES rb 1
MON_ITEM    rb 1
MON_MOVES   rb NUM_MOVES
MON_ID      rw 1
...
```

would go along with this:

```
box_struct: MACRO
\1Species:: db
\1Item::    db
\1Moves::   ds NUM_MOVES
\1ID::      dw
...
ENDM
```

For this issue, I think that the numerous local jumptables would be a good trial case, such as the `POKEGEARSTATE_*` constants for `PokegearJumptable.Jumptable`. Local jumptables are a common case that don't even all have constants yet anyway; the current practice of declaring their constants at the top of the file is awkward (it used to be necessary since some code above the jumptables would be using the constants); and the various jumptable constants would only get used in their own file, so there'd be no confusion about seeing `FOO` used in engine/bar.asm and wondering why there's no constants/foo_constants.asm file. (Whereas the `MON_*` constants are used in many places, so no matter whether they're defined with `EQUS`, `rsset`, or what, it's convenient to have them in constants/ if not in macros/.)

How do those sound?

**Comment by aaaaaa123456789** [comment]
I just opened [a feature request on RGBDS](https://github.com/gbdev/rgbds/issues/617) to make using these directives for offset definitions a lot cleaner. (Those dummy 1s in almost all the entries just look strange to me.)

As for the jumptables, I'm definitely not a fan of the `const_dw` name, but the idea seems fine.

**Comment by Rangi42** [comment]
I'd be happy with a better suggestion than `const_dw`, as long as it's clear that it's both defining a `const` and declaring a `dw`. (And ideally open to expansion, like if we later want a similar macro that declares a `dba` or `dab`.)

Or we could just interleave the existing macros:

```
.Jumptable:
	const_def 1
	const PINK_PAGE
	dw LoadPinkPage
	const GREEN_PAGE
	dw LoadGreenPage
	const BLUE_PAGE
	dw LoadBluePage
NUM_STAT_PAGES EQU const_value - 1
```

**Comment by aaaaaa123456789** [comment]
Perhaps the solution to the declaration problem is to include the declaration itself in the macro.

```
entry: MACRO
  const \1
  if _NARG == 2
    \2
  elif _NARG == 3
    \2, \3
  elif _NARG == 4
    \2, \3, \4
  else
    fail "Please extend this macro if necessary"
  endc
ENDM
```

Usage:

```
  const_def 1
  entry PINK_PAGE, dw LoadPinkPage
  entry GREEN_PAGE, dw LoadGreenPage
  entry BLUE_PAGE, dw LoadBluePage
```

**Comment by Rangi42** [comment]
Maybe so. Having the `dw` or whatever be part of the `entry` argument looks weird to me, not sure if it's better than having separate `const` and `dw` lines.

**Comment by vulcandth** [comment]
```asm
MACRO def_jump_table
\1:
	const_def
ENDM

MACRO jumpindex
	const \1
	dw \2
ENDM

PokegearJumptable:
	jumptable .Jumptable, wJumptableIndex

	def_jump_table .Jumptable
	jumpindex POKEGEARSTATE_CLOCKINIT,       PokegearClock_Init
	jumpindex POKEGEARSTATE_CLOCKJOYPAD,     PokegearClock_Joypad
	jumpindex POKEGEARSTATE_MAPCHECKREGION,  PokegearMap_CheckRegion
	jumpindex POKEGEARSTATE_JOHTOMAPINIT,    PokegearMap_Init
	jumpindex POKEGEARSTATE_JOHTOMAPJOYPAD,  PokegearMap_JohtoMap
	jumpindex POKEGEARSTATE_KANTOMAPINIT,    PokegearMap_Init
	jumpindex POKEGEARSTATE_KANTOMAPJOYPAD,  PokegearMap_KantoMap
	jumpindex POKEGEARSTATE_PHONEINIT,       PokegearPhone_Init
	jumpindex POKEGEARSTATE_PHONEJOYPAD,     PokegearPhone_Joypad
	jumpindex POKEGEARSTATE_MAKEPHONECALL,   PokegearPhone_MakePhoneCall
	jumpindex POKEGEARSTATE_FINISHPHONECALL, PokegearPhone_FinishPhoneCall
	jumpindex POKEGEARSTATE_RADIOINIT,       PokegearRadio_Init
	jumpindex POKEGEARSTATE_RADIOJOYPAD,     PokegearRadio_Joypad
```

Thought we might could do something like this for all the jumptables. The macro names can be changed to whatever... just placeholder names for now. All jump tables are inferred to be `dw` in size. I know we have idea's for other things (`dba` for example...) but for now, would something like this be okay for all the jumptables? If so, I can open a PR for it.

**Comment by Rangi42** [comment]
@mid-kid I'd like #978 to stay focused on scene scripts, since they're rather exceptional constants that are specific to maps, like the object event constants. Anyway, you said:

> Another place in the code that could use defining constants along with the pointers is the sprite anim and battle sprite anim subsystems, having to go back and forth between the constants and the tables to add entries is a major pain.

I'm not sure about this. Right now battle_anim_constants.asm has these series:

```
; BattleAnimObjects indexes (see data/battle_anims/objects.asm)
	const_def
	const ANIM_OBJ_HIT_BIG_YFIX
	...
	const ANIM_OBJ_PLAYERHEAD_2ROW
DEF NUM_ANIM_OBJS EQU const_value

; DoBattleAnimFrame arguments (see engine/battle_anims/functions.asm)
	const_def
	const BATTLEANIMFUNC_NULL
	...
	const BATTLEANIMFUNC_COTTON
DEF NUM_BATTLEANIMFUNCS EQU const_value

; BattleAnimFrameData indexes (see data/battle_anims/framesets.asm)
	const_def
	const BATTLEANIMFRAMESET_00
	...
	const BATTLEANIMFRAMESET_B8
DEF NUM_BATTLEANIMFRAMESETS EQU const_value

; BattleAnimOAMData indexes (see data/battle_anims/oam.asm)
	const_def
	const BATTLEANIMOAMSET_00
	...
	const BATTLEANIMOAMSET_D7
DEF NUM_BATTLEANIMOAMSETS EQU const_value

; BattleBGEffects indexes (see engine/battle_anims/bg_effects.asm)
	const_def 1
	const ANIM_BG_FLASH_INVERTED
	...
	const ANIM_BG_WOBBLE_SCREEN
DEF NUM_ANIM_BGS EQU const_value - 1

; wBattleAnimTileDict keys (see wram.asm)
; AnimObjGFX indexes (see data/battle_anims/object_gfx.asm)
	const_def 1
	const ANIM_GFX_HIT
	...
	const ANIM_GFX_ENEMYFEET
DEF NUM_ANIM_GFX EQU const_value - 1
```

It's a convenient place to have comments on what data tables or engine jumptables they correspond to. And if you see the constants being used in data/moves/animations.asm, this is the obvious file to check for them. Yes, they mostly only correspond to one data table each; but so do the `ICON_*` constants for `IconPointers`, but I don't think we should alter those. Also, for the `BATTLEANIMFUNC_*` and `ANIM_BG_*` constants that only correspond to an engine jumptable, I think they'd be harder to discover inside an engine file, versus a data or constant file.

**Comment by mid-kid** [comment]
> And having all these related constant series defined in one file, lets you read the comments and see all the data+engine files that might need updating when adding some completely new battle anim.

Often enough you don't need to add an entry in every single table to make a new battle anim object. In fact, when you have to, having to modify 4 * 2 tables instead of just 4 is what makes it so tedious.

I know that having all constants in constants/ aids for discoverability, but they make editing a true mightmare. More than once have I lost track of the order of these tables before we had comments for each entry, and now we do, we might just as well turn it into a comment.

Adapting other tables depends in my opinion on how much the constants are used across the codebase. I'm just saying that this is a pretty clear case where I'd just cut the verbosity and define the constants together with the table, since these constants are exclusively referenced from one other file, except for some framesets in engine/battle_anims/functions.asm.
